### PR TITLE
[ECO-5163] fix: duplicated messages because of duplicated attach message

### PR DESCRIPTION
**Pull request by ttypic**
Resolves https://github.com/ably/ably-java/issues/1050

Implemented missing [(RTL4i)](https://sdk.ably.com/builds/ably/specification/main/features/#RTL4i) spec point to prevent duplicated attach message

## Summary by CodeRabbit

- **New Features**
	- Added a conditional check to enhance channel attachment logic, preventing premature attach requests during specific connection states.
	- Introduced a new test to ensure message integrity by verifying that messages are not duplicated under certain subscription conditions.

- **Bug Fixes**
	- Improved message handling to prevent duplication of attach messages during connection events.

- **Tests**
	- Added a test method to validate message delivery without duplication.
	- Enhanced the flexibility of connection failure tests by using dynamic key assignments.

### CODE REVIEW

**Review by sacOO7 — COMMENTED** [review]
Thanks for raising the PR to fix the issue!
checks for `ATTACH` are bit tricky and has corner cases mentioned here https://github.com/ably/ably-java/issues/1018
So, I will take my time to carefully review the PR to make sure all cases are covered

**Comment by sacOO7** [comment]
Also, it seems some of the realtime tests failing

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java [hunk 1]
```diff
@@ -1010,4 +1014,42 @@ public void should_have_serial_action_createdAt() throws AblyException {
...
+                .collect(Collectors.toList());
+            channel.publish(messages.toArray(Message[]::new));
+        }
+
+        try (AblyRealtime realtime = new AblyRealtime(opts)) {
+            final ChannelOptions options = new ChannelOptions();
+            options.params = Map.of("rewind", "10");
+            final Channel channel = realtime.channels.get(testChannelName, options);
+            final CompletionWaiter completionWaiter = new CompletionWaiter();
+            final AtomicInteger counter = new AtomicInteger();
```
Comment (sacOO7) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1038]: Also, we can avoid using `AtomicInteger` since the internal `emit` call for channel messages is blocking in nature and the next event will only be sent when the previous one is processed

Comment (sacOO7) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1038]: Seems, IDE suggests to make it either atomic or final : (

Comment (sacOO7) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1038]: I think test can be further simplified to 
```
        ClientOptions opts = createOptions(testVars.keys[0].keyStr);
        opts.logLevel = 2;
        String testChannelName = "my-channel" + System.currentTimeMillis();
        try (AblyRest rest = new AblyRest(opts)) {
            final io.ably.lib.rest.Channel channel = rest.channels.get(testChannelName);
            Message[] messages = new Message[3];
            for (int i = 0; i < messages.length; i++) {
                messages[i] = new Message("name", "message" + i);
            }
            channel.publish(messages);
        }

        try (AblyRealtime realtime = new AblyRealtime(opts)) {
            final ChannelOptions options = new ChannelOptions();
            options.params = Map.of("rewind", "10");
            final Channel channel = realtime.channels.get(testChannelName, options);

            final int[] receivedMessages = {0};
            channel.subscribe((Channel.MessageListener) message -> receivedMessages[0] = receivedMessages[0] + 1);

            Exception conditionError = new Helpers.ConditionalWaiter().wait(() -> receivedMessages[0] == 3, 10000);
            assertNull(conditionError);

            assertEquals("Should be exactly 3 messages",3, receivedMessages[0]);

            Thread.sleep(1500);

            assertEquals("Should be exactly 3 messages even after 1.5 sec wait",3, receivedMessages[0]);
```
`ConditionalWaiter` is same as `assertWaiter`

Comment (ttypic) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1038]: it's a test, we just need wrapper around counter :) we don't care about atomic's overhead here (it's probably less than millisecond). Overhead from our `CompletionWaiter` are incomparable to it

Comment (sacOO7) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1038]: IDE is saying, `messages.toArray(Message[]::new)` is only compatible with java 11+

Comment (ttypic) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1037]: once again it's a test, this code will be never executed by library users we still target 1.8

Comment (sacOO7) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1037]: Yeah, but not sure why IDE is signaling warning for use of `List.of` and `Map.of` ( java 9+ compatible )

Comment (sacOO7) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1037]: I think we can ignore those warnings, but shouldn't change IDE settings for the same.
Doing so can introduce java 8+ code in the source code as well.

Comment (ttypic) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeMessageTest.java:1038]: I rewrote to use 1.8, to get rid of IDE warnings

**Review by sacOO7 — COMMENTED** [review]
Okay, not sure if this is a sandbox issue, but it seems `realtime.RealtimeMessageTest` are consistently failing on both `okhttp` and `websocket` connection.
You can check what's going wrong there : (

**Comment by sacOO7** [comment]
If this is a sandbox issue, then I will approve PR right away ( since `should_not_duplicate_messages` seems to be passing )

**Review by sacOO7 — APPROVED** [review]
LGTM
